trulens_eval/trulens_eval/feedback.py

```diff
@@ -1,5 +1,353 @@
+later section.
+
+### Global import requirement
+
+If using deferred evaluation, the feedback function implementations and
+aggregation implementations must be methods from a class that is globally
+importable. That is, the callables must be accessible were you to evaluate this
+code:
+
+```python
+import somepackage.[...].someclass 
+# [...] means optionally further package specifications
+
+provider = someclass(...) # constructor arguments can be included
+feedback_implementation = provider.somemethod
+```
+
+For provided feedback functions, `somepackage` is `trulens_eval.feedback` and
+`someclass` is `OpenAI` or one of the other `Provider` subclasses. Custom
+feedback functions likewise need to belong to a package that can be imported.
+Critically, classes defined locally in a notebook will not be importable this
+way.
+
+## Specifying Arguments
+
+The mapping between app/records to feedback implementation arguments is
+specified by the `on...` methods of the `Feedback` objects. The general form is:
+
+```python
+feedback: Feedback = feedback.on(argname1=selector1, argname2=selector2, ...)
+```
+
+That is, `Feedback.on(...)` returns a new `Feedback` object with additional
+argument mappings, the source of `argname1` is `selector1` and so on for further
+argument names. The types of `selector1` is `JSONPath` which we elaborate on in
+the further sections.
+
+If argument names are ommitted, they are taken from the feedback function
+implementation signature in order. That is, 
+
+```python
+...on(argname1=selector1, argname2=selector2)
+```
+
+and
+
+```python
+...on(selector1, selector2)
+```
+
+are equivalent assuming the feedback implementation has two arguments,
+`argname1` and `argname2`, in that order.
+
+Several utility methods starting with `.on` provide shorthands:
+
+- `on_input(arg) == on_prompt(arg: Optional[str])` -- both specify that the next
+    unspecified argument or `arg` should be the main app input.
+
+- `on_output(arg) == on_response(arg: Optional[str])` -- specify that the next
+    argument or `arg` should be the main app output.
+
+- `on_input_output() == on_input().on_output()` -- specifies that the first
+    two arguments of implementation should be the main app input and main app
+    output, respectively.
+
+- `on_default()` -- depending on signature of implementation uses either
+    `on_output()` if it has a single argument, or `on_input_output` if it has
+    two arguments.
+
+Some wrappers include additional shorthands:
+
+### llama_index-specific selectors
+
+- `TruLlama.select_source_nodes()` -- outputs the selector of the source
+    documents part of the engine output.
+
+### Running Feedback
+
+Feedback implementations are simple callables that can be run on any arguments
+matching their signatures. However, once wrapped with `Feedback`, they are meant
+to be run on outputs of app evaluation (the "Records"). Specifically,
+`Feedback.run` has this definition:
```

Comment:
Maybe move this to the top, under **Feedback implementations**

Reply:
moved part